### PR TITLE
Add proper validation for empty batch in External Source

### PR DESCRIPTION
**Pull request by klecki**
Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

## Description:
The Pipeline doesn't work with empty inputs and errors out when one tries to return them.
As the only place we can get empty input from (for now) is External Source, instead of erroring
in the executor we now check the input batch and error out early.

Previous error:
```
Error when executing Mixed operator MakeContiguous encountered:
[/home/klecki/work/salvador/dali/dali/pipeline/data/tensor_list.cc:519] Assert on "IsValidType(new_type)" failed: TensorList cannot be resized with invalid type. To zero out the TensorList Reset() can be used.
```

New error:
```
RuntimeError: [/home/klecki/work/salvador/dali/dali/pipeline/operator/builtin/external_source.h:496] Assert on "batch.num_samples() > 0" failed: ExternalSource expects non-empty batches to be provided as
the input. Got batch with 0 samples.
```

### Affected modules and functionalities:
ES

### Key points relevant for the review:
N/A

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Review by JanuszL — APPROVED** [review]

**Review by szalpal — APPROVED** [review]

**Comment by klecki** [comment]
!build

File: dali/test/python/test_external_source_dali.py [hunk 1]
```diff
@@ -16,10 +16,11 @@
 import nvidia.dali.fn as fn
 import nvidia.dali.ops as ops
 import nvidia.dali.types as types
+import nvidia.dali.tensors as tensors
```
Comment (klecki) [inline dali/test/python/test_external_source_dali.py:19]:
```suggestion
```

**Comment by klecki** [comment]
!build

File: dali/test/python/test_external_source_dali.py [hunk 2]
```diff
@@ -562,3 +562,20 @@ def test_non_utilized_external_source_pruning():
...
+        return fn.external_source(source=lambda: [])
+
+    # Providing an empty batch was legal, but it failed on MakeContiguous node.
+    # This checks proper validation in External Source which is the only way that could provide
+    # empty batch as input into DALI graph.
+    with assert_raises(RuntimeError, glob="*ExternalSource expects non-empty batches*"):
+        pipe = pipeline(batch_size=max_batch_size, num_threads=4, device_id=0)
+        pipe.build()
+        pipe.run()
+
```
Comment (klecki) [inline dali/test/python/test_external_source_dali.py:581]:
```suggestion
```

**Comment by klecki** [comment]
!build

**Pull request merged**